File: client/ruby-hijiki/lib/hijiki/dcmgr_resource/12.03/network.rb
# -*- coding: utf-8 -*-
module Hijiki::DcmgrResource::V1203

  class Network < Base
    initialize_user_result nil, [:account_id,
                                 :uuid,
                                 :ipv4_network,
                                 :ipv4_gw,
                                 :prefix,
                                 :metric,
                                 :domain_name,
                                 :dns_server,
                                 :dhcp_server,
                                 :metadata_server,
                                 :metadata_server_port,
                                 # :nat_network_id,
                                 # :dc_network_id,
                                 :description,
                                 :created_at,
                                 :updated_at,
                                 # :gateway_network_id,
                                 :network_mode,
                                 :bandwidth,
                                 :service_type,
                                 :display_name,
                                 # :bandwidth_mark,
                                ], [:dc_network], [:network_services]
    initialize_user_result 'DcNetwork', [:uuid,
                                         :name,
                                         :description,
                                         :offering_network_modes,
                                         :created_at,
                                         :updated_at,
                                        ]
    initialize_user_result 'NetworkService', [:name,
                                              :description,
                                              :created_at,
                                              :updated_at,
                                             ]

    module ClassMethods
      include Hijiki::DcmgrResource::Common::ListMethods::ClassMethods

      def list(params = {})
        super(params.merge({:state=>'alive_with_terminated'}))
      end
      
      def create(params)
        object = self.new
        object.display_name = params[:display_name] if params[:display_name]
        object.description = params[:description] if params[:description]
        object.domain_name = params[:domain_name] if params[:domain_name]
        object.dc_network = params[:dc_network] if params[:dc_network]
        object.network_mode = params[:network_mode]
        object.network = params[:ipv4_network]
        object.gw = params[:ipv4_gw] if params[:ipv4_gw]
        object.prefix = params[:prefix]
        object.ip_assignment = params[:ip_assignment] if params[:ip_assignment]
        object.editable = params[:editable] if params[:editable]
        
        object.service_dhcp = params[:service_dhcp] if params[:service_dhcp]
        object.service_dns = params[:service_dns] if params[:service_dns]
        object.service_gateway = params[:service_gateway] if params[:service_gateway]

        object.dhcp_range = params[:dhcp_range] if params[:dhcp_range]

        object.save
        object
      end
    end
    extend ClassMethods

    def find_vif(vif_id)
      NetworkVif.find(vif_id, :params => { :network_id => self.id })
    end

    def find_service(service_name)
      NetworkService.find(service_name, :params => { :network_id => self.id })
    end

    def list_services
      NetworkService.list(:network_id => self.id)
    end

    def delete_service(vif_id, name)
      self.delete('services', {:vif_id => vif_id, :name => name})
    end

    def get_dhcp_ranges
      self.get(:dhcp_ranges)
    end

    def add_dhcp_range(range_begin, range_end)
      self.put('dhcp_ranges/add', { :range_begin => range_begin, :range_end => range_end })
    end

    def remove_dhcp_range(range_begin, range_end)
      self.put('dhcp_ranges/remove', { :range_begin => range_begin, :range_end => range_end })
    end

    def attach(vif_id)
      self.put("vifs/#{vif_id}/attach")
    end

    def detach(vif_id)
      self.put("vifs/#{vif_id}/detach")
    end
  end

end

File: client/go-wakamevdc/network.go
package wakamevdc

import (
	"fmt"
	"net/http"
)

const NetworkPath = "networks"

type Network struct {
	ID                 string   `json:"id"`
	AccountID          string   `json:"account_id"`
	IPv4Network        string   `json:"ipv4_network"`
	IPv4GW             string   `json:"ipv4_gw"`
	Prefix             int      `json:"prefix"`
	NATNetworkID       string   `json:"nat_network_id"`
	Metric             int      `json:"metric"`
	IPAssignment       string   `json:"ip_assignment"`
	Editable           bool     `json:"editable"`
	DomainName         string   `json:"domain_name"`
	DNSServer          string   `json:"dns_server"`
	DHCPServer         string   `json:"dhcp_server"`
	MetadataServer     string   `json:"metadata_server"`
	MetadataServerPort int      `json:"metadata_server_port"`
	NetworkMode        string   `json:"network_mode"`
	NetworkServices    []string `json:"network_services"`
	DCNetwork          struct {
		ID                   string   `json:"id"`
		Name                 string   `json:"name"`
		Description          string   `json:"description"`
		VLANLeaseID          string   `json:"vlan_lease_id"`
		OfferringNetworkMode []string `json:"offering_network_modes"`
		AllowNewNetworks     bool     `json:"allow_new_networks"`
		CreatedAt            string   `json:"created_at"`
		UpdatedAt            string   `json:"updated_at"`
	} `json:"dc_network"`
	ServiceType string `json:"service_type"`
	DisplayName string `json:"display_name"`
	Description string `json:"description"`
	CreatedAt   string `json:"created_at"`
	UpdatedAt   string `json:"updated_at"`
	DeletedAt   string `json:"deleted_at"`
}

type NetworkService struct {
	client *Client
}

type NetworkCreateParams struct {
	AccountID    string `url:"account_id,omitempty"`
	IPv4Network  string `url:"network"`
	IPv4GW       string `url:"gw,omitempty"`
	Prefix       int    `url:"prefix"`
	DCNetworkID  string `url:"dc_network"`
	NetworkMode  string `url:"network_mode,omitempty"`
	DomainName   string `url:"domain_name,omitempty"`
	IPAssignment string `url:"ip_assignment,omitempty"`
	Editable     bool   `url:"editable,omitempty"`
	Metric       int    `url:"metric,omitempty"`
	ServiceType  string `url:"service_type,omitempty"`
	Description  string `url:"description,omitempty"`
	DisplayName  string `url:"display_name,omitempty"`
	// TODO: service_dhcp, service_dns, service_gateway
}

type NetworkUpdateParams struct {
	DisplayName string `url:"display_name,omitempty"`
}

func (s *NetworkService) Create(req *NetworkCreateParams) (*Network, *http.Response, error) {
	nw := new(Network)
	resp, err := trapAPIError(func(errResp *ErrorResponse) (*http.Response, error) {
		return s.client.Sling().Post(NetworkPath).BodyForm(req).Receive(nw, errResp)
	})

	return nw, resp, err
}

func (s *NetworkService) Update(id string, req *NetworkUpdateParams) (*http.Response, error) {
	resp, err := trapAPIError(func(errResp *ErrorResponse) (*http.Response, error) {
		return s.client.Sling().Put(fmt.Sprintf(NetworkPath+"/%s", id)).BodyForm(req).Receive(nil, errResp)
	})
	return resp, err
}

func (s *NetworkService) Delete(id string) (*http.Response, error) {
	return trapAPIError(func(errResp *ErrorResponse) (*http.Response, error) {
		return s.client.Sling().Delete(fmt.Sprintf(NetworkPath+"/%s", id)).Receive(nil, errResp)
	})
}

func (s *NetworkService) GetByID(id string) (*Network, *http.Response, error) {
	nw := new(Network)
	resp, err := trapAPIError(func(errResp *ErrorResponse) (*http.Response, error) {
		return s.client.Sling().Get(fmt.Sprintf(NetworkPath+"/%s", id)).Receive(nw, errResp)
	})

	return nw, resp, err
}

type NetworksList struct {
	Total   int       `json:"total"`
	Start   int       `json:"start"`
	Limit   int       `json:"limit"`
	Results []Network `json:"results"`
}

func (s *NetworkService) List(req *ListRequestParams) (*NetworksList, *http.Response, error) {
	nwList := make([]NetworksList, 1)
	resp, err := trapAPIError(func(errResp *ErrorResponse) (*http.Response, error) {
		return s.client.Sling().Get(NetworkPath).QueryStruct(req).Receive(&nwList, errResp)
	})

	if err == nil && len(nwList) > 0 {
		return &nwList[0], resp, err
	}
	// Return empty list object.
	return &NetworksList{}, resp, err
}

func (s *NetworkService) DHCPRangeList(id string) ([][]string, *http.Response, error) {
	var drl [][]string

	resp, err := trapAPIError(func(errResp *ErrorResponse) (*http.Response, error) {
		return s.client.Sling().Get(fmt.Sprintf(NetworkPath+"/%s/dhcp_ranges", id)).Receive(&drl, errResp)
	})

	return drl, resp, err
}

type DHCPRangeCreateParams struct {
	RangeBegin string `url:"range_begin"`
	RangeEnd   string `url:"range_end"`
}

func (s *NetworkService) DHCPRangeCreate(id string, req *DHCPRangeCreateParams) (*http.Response, error) {
	resp, err := trapAPIError(func(errResp *ErrorResponse) (*http.Response, error) {
		return s.client.Sling().Put(fmt.Sprintf(NetworkPath+"/%s/dhcp_ranges/add", id)).BodyForm(req).Receive(nil, errResp)
	})

	return resp, err
}

type DHCPRangeDeleteParams struct {
	RangeBegin string `url:"range_begin"`
	RangeEnd   string `url:"range_end"`
}

func (s *NetworkService) DHCPRangeDelete(id string, req *DHCPRangeDeleteParams) (*http.Response, error) {
	resp, err := trapAPIError(func(errResp *ErrorResponse) (*http.Response, error) {
		return s.client.Sling().Put(fmt.Sprintf(NetworkPath+"/%s/dhcp_ranges/remove", id)).BodyForm(req).Receive(nil, errResp)
	})

	return resp, err
}
